python/rmm/_lib/device_buffer.pyx

```diff
@@ -71,13 +75,8 @@ cdef class DeviceBuffer:
             else:
                 self.c_obj.reset(new device_buffer(c_ptr, size, c_stream))
 
-                if c_stream == NULL:
-                    err = cudaStreamSynchronize(c_stream)
-                    if err != cudaError.cudaSuccess:
-                        with gil:
-                            raise RuntimeError(
-                                f"Stream sync failed with error: {err}"
-                            )
+            if c_stream.is_default():
+                c_stream.synchronize()
```

Comment:
Starting thread for discussion. Do we want to synchronize whenever the stream is:
- the stream is the default stream (could be either legacy default or per thread default)
- the stream is explicitly the legacy default stream
- the stream is the default stream only when it refers to the legacy default stream (i.e. `CudaStreamView(<cudaStream_t>0)` and per thread default stream is **not** enabled
- the stream is not specified by the user (we could either use `None` as a placeholder in this situation, or create a new stream persistently that we specifically check for)

I'm in favor of the last option in using a separate CUDA stream as the default parameter, and when this separate CUDA stream is used we synchronize.

Something like this:
```
synchronized_stream = cuda_stream()
synchronized_stream_view = CudaStreamView(synchronized_stream)

cdef class DeviceBuffer:
    def __cinit__(self, *, uintptr_t ptr=0, size_t size=0, CudaStreamView stream=synchronized_stream_view):
        ...
```

Reply:
The challenge with this approach is we'd need to create a cuda stream at library import / load time, which means that the driver and/or runtime will be initialized and would break something like Dask trying to shuffle `CUDA_VISIBLE_DEVICES`.

I guess we could create some "lazy" wrappers that don't actually create the cuda stream until the attribute is accessed and then memoize it into the instance.

Reply:
Well the default stream bit may be a red-herring. The issue is the data we are passed may expire or be altered before the copy is finished. The issue is shown in this comment ( https://github.com/rapidsai/rmm/issues/313#issuecomment-591840902 ). So we need some way of ensuring the copy  is completed. Currently we do that with synchronizing. Maybe there are other alternatives? Though the condition probably shouldn't be here.

Reply:
Yes, by default we should protect against these things, but if a user explicitly gives us a stream then it's them basically opting into the asynchronous programming model. In this programming model, they're required to make sure the object doesn't go out of scope or get modified before the operation finishes.

Reply:
Maybe we should consider adding a flag to the constructor for disabling synchronization

Reply:
We'd need to add that flag to every method that takes a stream, which is presumably every method.

Reply:
Couple things: I'm opposed to the idea of creating an explicit stream -- the problem with CUDA today is it has a default stream with different semantics from non-default streams. I don't want to create a third classification.

Second, you don't need to provide a flag to every method that takes a stream. You just need synchronous and asynchronous methods with different names: For example, see `device_uvector.hpp`: `device_uvector::set_element_async()` takes a stream, and does a `cudaMemcpyAsync` on that stream, but does not synchronize the stream. `  `device_uvector::set_element()` also takes a stream, and does a `cudaMemcpyAsync` on that stream, but does NOT synchronize the stream.

Reply:
@kkraus14 this PR doesn't change the previous semantics of `DeviceBuffer`. I propose that we discuss and alter the semantics in a separate PR if that's OK.

Reply:
I think the only challenge is that because we're moving away from using an `int` to using a Cython extension type, it changes the GIL semantics of these functions. It was very nice that previously these functions were callable / usable without the GIL.

Reply:
Resolving this and we'll revisit in a follow up PR